### PR TITLE
<feat>(kktix): added gather town whitelist feature

### PR DESCRIPTION
**Pull request by david30907d**
## Types of changes

- **New feature**

## Description

Once user bought a ticket on KKTIX, Airflow would ingest these data every 5 mins and whitelist his/her email into gathertown automatically

### TODO

this PR doesn't handle the edge case, which is users refund their ticket but still keep his/her gathertown login permission. We can punt on this for now until wrapping up all the necessities for Conference.
cc @henry410213028 @GeekAngus 

## Checklist:
- [ ] Update the documentation if necessary
- [ ] Need to figure out how to handle `refunded ticket` behavior...
## Steps to Test This Pull Request

TBD 
## Expected behavior

should see your nickname after logging into Gathertown, however right now you can only see `anonymous`

## Additional context

https://hackmd.io/PM_sWO5USo6dxMqT1uCrCQ?view#Solution and check `gather town section`

### CODE REVIEW

File: dags/ods/kktix_ticket_orders/udfs/gather_town_loader.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
+"""
+load user's name, email etc into gather town whitelist
+please refer to this document for details: https://hackmd.io/PM_sWO5USo6dxMqT1uCrCQ?view
+"""
```
Comment (david30907d) [inline dags/ods/kktix_ticket_orders/udfs/gather_town_loader.py:4]: gather town logic

File: dags/ods/kktix_ticket_orders/udfs/gather_town_loader.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+GATHERTOWN_HTTP_HOOK = HttpHook(http_conn_id="gathertown_api", method="POST")
+
+
+def load(**context):
+    event_raw_data_array = context["ti"].xcom_pull(task_ids="GET_ATTENDEE_INFOS")
+    for event_raw_data in event_raw_data_array[:1]:
+        resp = GATHERTOWN_HTTP_HOOK.run_with_advanced_retry(
+            endpoint="/api/setEmailGuestlist",
```
Comment (david30907d) [inline dags/ods/kktix_ticket_orders/udfs/gather_town_loader.py:23]: main logic is as follow:
1. get sensitive data like email and name from Airflow's xcom
2. invoke http post (provided by @GeekAngus 🙏 )

File: dags/ods/kktix_ticket_orders/kktix_dag.py [hunk 3]
```diff
@@ -34,7 +34,13 @@
...
     )
 
-    CREATE_TABLE_IF_NEEDED >> GET_ATTENDEE_INFOS
+    ADD_USER_TO_GATHER_TOWN_WHITELIST = PythonOperator(
+        task_id="ADD_USER_TO_GATHER_TOWN_WHITELIST",
+        python_callable=gather_town_loader.load,
+        provide_context=True,
+    )
+
+    CREATE_TABLE_IF_NEEDED >> GET_ATTENDEE_INFOS >> ADD_USER_TO_GATHER_TOWN_WHITELIST
```
Comment (david30907d) [inline dags/ods/kktix_ticket_orders/kktix_dag.py:43]: task `ADD_USER_TO_GATHER_TOWN_WHITELIST` depends on `GET_ATTENDEE_INFOS`

also, `GET_ATTENDEE_INFOS` would pass a payload via `XCOM` to `ADD_USER_TO_GATHER_TOWN_WHITELIST`

File: dags/ods/kktix_ticket_orders/udfs/kktix_api.py [hunk 4]
```diff
@@ -25,13 +25,19 @@ def main(**context):
...
+        timestamp=timestamp,
+    )
     # load name and email to mailer before data has been hashed
     klaviyo_loader.load(event_raw_data_array)
     transformed_event_raw_data_array = kktix_transformer.transform(event_raw_data_array)
     kktix_loader.load(transformed_event_raw_data_array)
     print(f"Loaded {len(transformed_event_raw_data_array)} rows to BigQuery!")
 
+    # pass these unhashed data through xcom to next airflow task
+    return kktix_transformer._extract_sensitive_unhashed_raw_data(event_raw_data_array)
```
Comment (david30907d) [inline dags/ods/kktix_ticket_orders/udfs/kktix_api.py:39]: simply use a `return` would pass a payload to next airflow task via `XCOM`

this is how `ADD_USER_TO_GATHER_TOWN_WHITELIST` get those sensitive data

File: dags/ods/kktix_ticket_orders/udfs/gather_town_loader.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            _retry_args=RETRY_ARGS,
+            json={
+                "spaceId": "VUJexktI2UhoGaun\\FunFunPython",
```
[the remark this answers is not in the document]
Comment (david30907d) [reply dags/ods/kktix_ticket_orders/udfs/gather_town_loader.py:26]: good catch, just realized `FunFunPython` is a staging Gather Town 😂

**Review by GeekAngus — APPROVED** [review]
it is great, thank you

**Pull request merged**